src/SetupPage.js

```diff
@@ -163,41 +163,59 @@ class PreviousEnvironments extends React.Component {
 
   render() {
     const { environments, loading } = this.state;
-    if (loading) {
-      return (
-        <div id="previous">
-          <h3>PreviousEnvironments</h3>
-          <span>...</span>
-        </div>
-      );
-    }
-
     return (
       <div id="previous">
         <h3>Previous Environments</h3>
-        <ul>
-          {environments.map(env => (
-            <li key={env.shortlink}>
-              <Link to={`/s/${env.shortlink}`}>
-                {env.owner}/{env.repo}
-              </Link>
-              <span className="names">
-                {env.revisions.map(r => r[0]).join(', ')}
-              </span>
-            </li>
-          ))}
-        </ul>
+        {loading ? (
+          <EllipsisLoading />
+        ) : (
+          <ul>
+            {environments.map(env => (
+              <li key={env.shortlink}>
+                <Link to={`/s/${env.shortlink}`}>
+                  {env.owner}/{env.repo}
+                </Link>
+                <span className="names">
+                  {env.revisions.map(r => r[0]).join(', ')}
+                </span>
+              </li>
+            ))}
+          </ul>
+        )}
       </div>
     );
   }
 }
 
+class EllipsisLoading extends React.PureComponent {
```

Comment:
I "stole" this rather big and convoluted code from my own `admin.peterbe.com` project.